[JSC] Add ChangePrototype transition

Constellation opened the pull request:
#### 4d1e380621d4df47c82344c0ca20d2162169d6a1
<pre>
[JSC] Add ChangePrototype transition
<a href="https://bugs.webkit.org/show_bug.cgi?id=277061">https://bugs.webkit.org/show_bug.cgi?id=277061</a>
<a href="https://rdar.apple.com/132453886">rdar://132453886</a>

Reviewed by Keith Miller.

This patch adds transition tracking for changing [[Prototype]]. Some code is frequently changing [[Prototype]] to create a class,
and this patch teaches the code to use the same Structure by tracing the transition chain to leverage this characteristics in IC and other optimizations.

* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
(JSC::InlineCacheCompiler::compileOneAccessCaseHandler):
* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::StructureTransitionTable::contains const):
(JSC::StructureTransitionTable::add):
(JSC::Structure::materializePropertyTable):
(JSC::Structure::changePrototypeTransition):
(JSC::dumpTransitionKind):
* Source/JavaScriptCore/runtime/StructureInlines.h:
(JSC::StructureTransitionTable::Hash::createFromStructure):
(JSC::StructureTransitionTable::get const):
* Source/JavaScriptCore/runtime/StructureTransitionTable.h:
(JSC::StructureTransitionTable::PointerKey::PointerKey):
(JSC::StructureTransitionTable::PointerKey::raw const):
(JSC::StructureTransitionTable::PointerKey::pointer const):
(JSC::StructureTransitionTable::PointerKey::fromRaw):
(JSC::StructureTransitionTable::Hash::Key::Key):
(JSC::StructureTransitionTable::Hash::Key::impl const):
(JSC::StructureTransitionTable::Hash::createKey):
(JSC::StructureTransitionTable::Hash::hash):

Canonical link: <a href="https://commits.webkit.org/281414@main">https://commits.webkit.org/281414@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a924f60042c06f8a24f15ec57eaa545da7136ba9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59838 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39185 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12368 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63755 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10362 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46838 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10535 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48530 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7250 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61868 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36552 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51815 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29374 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33259 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9053 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9285 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/52932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55187 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9331 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65485 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/59083 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3766 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/9199 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/55870 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3777 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51793 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56011 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3129 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/80841 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8956 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34997 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14031 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36080 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37166 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/35825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/runtime/Structure.cpp:
```diff
@@ -672,15 +675,34 @@ Structure* Structure::changePrototypeTransition(VM& vm, Structure* structure, JS
 
+    // Changing [[Prototype]] means that we refresh this object completely.
+    // This is very likely that this object will behaves differently from the previous one.
+    // Let's pin the table and berak the edge to the previous Structure.
```
Typo: Break

Constellation replied:
Fixed.

kmiller68 commented on Source/JavaScriptCore/runtime/Structure.cpp:
```diff
@@ -52,7 +52,7 @@ inline void StructureTransitionTable::setSingleTransition(VM& vm, JSCell* owner,
 }
 
-bool StructureTransitionTable::contains(UniquedStringImpl* rep, unsigned attributes, TransitionKind transitionKind) const
+bool StructureTransitionTable::contains(void* rep, unsigned attributes, TransitionKind transitionKind) const
```
I think it would be better to have a union type here for readability.

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/runtime/StructureTransitionTable.h:
```diff
@@ -164,12 +187,12 @@ class StructureTransitionTable {
+            Key(PointerKey impl, unsigned attributes, TransitionKind transitionKind)
+                : m_encodedData(impl.raw() | (static_cast<uintptr_t>(attributes) << attributesShift) | (static_cast<uintptr_t>(transitionKind) << transitionKindShift))
             {
                 ASSERT(impl == this->impl());
```
Can we do `ASSERT(roundUpToMultipleOf<8>(impl.raw()) == impl.raw())` here?

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/runtime/Structure.cpp:
```diff
@@ -672,15 +675,34 @@ Structure* Structure::changePrototypeTransition(VM& vm, Structure* structure, JS
+    if (shouldChain) {
+        ASSERT(structure->isObject());
+        if (Structure* existingTransition = structure->m_transitionTable.get(key, 0, TransitionKind::ChangePrototype)) {
+            existingTransition->checkOffsetConsistency();
```
Can we add a `ASSERT(!existingTransition->hasPolyProto());` here?

Constellation replied:
Done.

kmiller68 reviewed (approved):
r=me with nits.

Constellation committed:
[JSC] Add ChangePrototype transition
https://bugs.webkit.org/show_bug.cgi?id=277061
rdar://132453886

Reviewed by Keith Miller.

This patch adds transition tracking for changing [[Prototype]]. Some code is frequently changing [[Prototype]] to create a class,
and this patch teaches the code to use the same Structure by tracing the transition chain to leverage this characteristics in IC and other optimizations.

* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
(JSC::InlineCacheCompiler::compileOneAccessCaseHandler):
* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::StructureTransitionTable::contains const):
(JSC::StructureTransitionTable::add):
(JSC::Structure::materializePropertyTable):
(JSC::Structure::changePrototypeTransition):
(JSC::dumpTransitionKind):
* Source/JavaScriptCore/runtime/StructureInlines.h:
(JSC::StructureTransitionTable::Hash::createFromStructure):
(JSC::StructureTransitionTable::get const):
* Source/JavaScriptCore/runtime/StructureTransitionTable.h:
(JSC::StructureTransitionTable::PointerKey::PointerKey):
(JSC::StructureTransitionTable::PointerKey::raw const):
(JSC::StructureTransitionTable::PointerKey::pointer const):
(JSC::StructureTransitionTable::PointerKey::fromRaw):
(JSC::StructureTransitionTable::Hash::Key::Key):
(JSC::StructureTransitionTable::Hash::Key::impl const):
(JSC::StructureTransitionTable::Hash::createKey):
(JSC::StructureTransitionTable::Hash::hash):

Canonical link: https://commits.webkit.org/281414@main
